File: tst/cv.c
#include "pico.h"
#include <stdio.h>

int main (void) {
    pico_init(1);
    pico_set_scene_tile((Pico_Abs_Dim){0, 0});
    Pico_Rel_Dim dim = { '!', {100, 100} };
    pico_set_scene_dim(dim);

    ///////////////////////////////////////////////////////////////////////////
    // _to / _from across named layers
    ///////////////////////////////////////////////////////////////////////////

    printf("=== pos_to/_from world <-> window ===\n");
    {
        puts("(50,50) world -> window");
        Pico_Rel_Pos w   = { '!', {50, 50}, PICO_ANCHOR_NW };
        Pico_Rel_Pos win = { '!', {0, 0}, PICO_ANCHOR_NW };
        pico_cv_pos("window", &win, NULL, &w);
        assert((int)win.x == 250 && (int)win.y == 250);
    }
    {
        puts("(250,250) window -> world");
        Pico_Rel_Pos win = { '!', {250, 250}, PICO_ANCHOR_NW };
        Pico_Rel_Pos w   = { '!', {0, 0}, PICO_ANCHOR_NW };
        pico_cv_pos(NULL, &w, "window", &win);
        assert((int)w.x == 50 && (int)w.y == 50);
    }
    {
        puts("round-trip world <-> window");
        Pico_Rel_Pos orig = { '!', {37, 89}, PICO_ANCHOR_NW };
        Pico_Rel_Pos win  = { '!', {0, 0}, PICO_ANCHOR_NW };
        Pico_Rel_Pos back = { '!', {0, 0}, PICO_ANCHOR_NW };
        pico_cv_pos("window", &win, NULL, &orig);
        pico_cv_pos(NULL, &back, "window", &win);
        assert((int)back.x == 37 && (int)back.y == 89);
    }

    printf("=== sub-layer 2-hop walk ===\n");
    pico_layer_empty("world", "sub_cv", 1, (Pico_Rel_Dim){'!', {50, 50}}, NULL);
    pico_set_layer("sub_cv");

    {
        puts("(10,20) sub -> (20,40) world");
        Pico_Rel_Pos s = { '!', {10, 20}, PICO_ANCHOR_NW };
        Pico_Rel_Pos w = { '!', {0, 0}, PICO_ANCHOR_NW };
        pico_cv_pos("world", &w, NULL, &s);
        assert((int)w.x == 20 && (int)w.y == 40);
    }
    {
        puts("(20,40) world -> (10,20) sub");
        Pico_Rel_Pos w = { '!', {20, 40}, PICO_ANCHOR_NW };
        Pico_Rel_Pos s = { '!', {0, 0}, PICO_ANCHOR_NW };
        pico_cv_pos(NULL, &s, "world", &w);
        assert((int)s.x == 10 && (int)s.y == 20);
    }
    {
        puts("(10,20) sub -> (100,200) window (2 hops)");
        Pico_Rel_Pos s   = { '!', {10, 20}, PICO_ANCHOR_NW };
        Pico_Rel_Pos win = { '!', {0, 0}, PICO_ANCHOR_NW };
        pico_cv_pos("window", &win, NULL, &s);
        assert((int)win.x == 100 && (int)win.y == 200);
    }
    {
        puts("(100,200) window -> (10,20) sub (2 hops)");
        Pico_Rel_Pos win = { '!', {100, 200}, PICO_ANCHOR_NW };
        Pico_Rel_Pos s   = { '!', {0, 0}, PICO_ANCHOR_NW };
        pico_cv_pos(NULL, &s, "window", &win);
        assert((int)s.x == 10 && (int)s.y == 20);
    }

    printf("=== rect_to / rect_from ===\n");
    {
        puts("rect (10,20,5,10) sub -> (20,40,10,20) world");
        Pico_Rel_Rect s = { '!', {10, 20, 5, 10}, PICO_ANCHOR_NW };
        Pico_Rel_Rect w = { '!', {0, 0, 0, 0}, PICO_ANCHOR_NW };
        pico_cv_rect("world", &w, NULL, &s);
        assert((int)w.x==20 && (int)w.y==40);
        assert((int)w.w==10 && (int)w.h==20);
    }
    {
        puts("rect (20,40,10,20) world -> (10,20,5,10) sub");
        Pico_Rel_Rect w = { '!', {20, 40, 10, 20}, PICO_ANCHOR_NW };
        Pico_Rel_Rect s = { '!', {0, 0, 0, 0}, PICO_ANCHOR_NW };
        pico_cv_rect(NULL, &s, "world", &w);
        assert((int)s.x==10 && (int)s.y==20);
        assert((int)s.w==5  && (int)s.h==10);
    }

    printf("=== dim_to / dim_from ===\n");
    {
        puts("dim (5,10) sub -> (10,20) world");
        Pico_Rel_Dim s = { '!', {5, 10} };
        Pico_Rel_Dim w = { '!', {0, 0} };
        pico_cv_dim("world", &w, NULL, &s);
        assert((int)w.w==10 && (int)w.h==20);
    }
    {
        puts("dim (10,20) world -> (5,10) sub");
        Pico_Rel_Dim w = { '!', {10, 20} };
        Pico_Rel_Dim s = { '!', {0, 0} };
        pico_cv_dim(NULL, &s, "world", &w);
        assert((int)s.w==5 && (int)s.h==10);
    }

    pico_set_layer("world");

    printf("=== descendant target/source (bidirectional) ===\n");
    {
        puts("(20,40) world -> (10,20) sub (target=descendant)");
        Pico_Rel_Pos w = { '!', {20, 40}, PICO_ANCHOR_NW };
        Pico_Rel_Pos s = { '!', {0, 0}, PICO_ANCHOR_NW };
        pico_cv_pos("sub_cv", &s, NULL, &w);
        assert((int)s.x == 10 && (int)s.y == 20);
    }
    {
        puts("(10,20) sub -> (20,40) world (source=descendant)");
        Pico_Rel_Pos s = { '!', {10, 20}, PICO_ANCHOR_NW };
        Pico_Rel_Pos w = { '!', {0, 0}, PICO_ANCHOR_NW };
        pico_cv_pos(NULL, &w, "sub_cv", &s);
        assert((int)w.x == 20 && (int)w.y == 40);
    }
    {
        puts("rect (20,40,10,20) world -> (10,20,5,10) sub (target=descendant)");
        Pico_Rel_Rect w = { '!', {20, 40, 10, 20}, PICO_ANCHOR_NW };
        Pico_Rel_Rect s = { '!', {0, 0, 0, 0}, PICO_ANCHOR_NW };
        pico_cv_rect("sub_cv", &s, NULL, &w);
        assert((int)s.x==10 && (int)s.y==20);
        assert((int)s.w==5  && (int)s.h==10);
    }
    {
        puts("dim (10,20) world -> (5,10) sub (target=descendant)");
        Pico_Rel_Dim w = { '!', {10, 20} };
        Pico_Rel_Dim s = { '!', {0, 0} };
        pico_cv_dim("sub_cv", &s, NULL, &w);
        assert((int)s.w==5 && (int)s.h==10);
    }

    printf("=== aspect-fill mid layer (cv bug repro) ===\n");
    // mid 50x50 under world (100x100); zero w/h in scene.dst aspect-fills
    // from mid's own dim. Eager-resolved in pico_set_scene_dst so cv walks
    // see a concrete '!' rect (bypassing the cv NULL-ratio blind spot).
    pico_layer_empty("world", "mid", 1, (Pico_Rel_Dim){'!', {50, 50}}, NULL);
    pico_set_layer("mid");

    {
        puts("'!' dst {0,0,0,100}: (10,10) mid -> (20,20) world");
        pico_set_scene_dst (
            (Pico_Rel_Rect){'!', {0, 0, 0, 100}, PICO_ANCHOR_NW}
        );
        Pico_Rel_Pos m = { '!', {10, 10}, PICO_ANCHOR_NW };
        Pico_Rel_Pos w = { '!', {0, 0}, PICO_ANCHOR_NW };
        pico_cv_pos("world", &w, NULL, &m);
        printf("  got (%g,%g)\n", w.x, w.y);
        assert((int)w.x == 20 && (int)w.y == 20);
    }
    {
        puts("'%%' dst {0,0,0,1.0}: same fill, parent-relative");
        pico_set_scene_dst (
            (Pico_Rel_Rect){'%', {0, 0, 0, 1.0}, PICO_ANCHOR_NW}
        );
        Pico_Rel_Pos m = { '!', {10, 10}, PICO_ANCHOR_NW };
        Pico_Rel_Pos w = { '!', {0, 0}, PICO_ANCHOR_NW };
        pico_cv_pos("world", &w, NULL, &m);
        printf("  got (%g,%g)\n", w.x, w.y);
        assert((int)w.x == 20 && (int)w.y == 20);
    }
    {
        puts("'#' dst {1,1,0,10} tile=10: same fill, tile-relative");
        pico_set_scene_tile((Pico_Abs_Dim){10, 10});
        pico_set_scene_dst (
            (Pico_Rel_Rect){'#', {1, 1, 0, 10}, PICO_ANCHOR_NW}
        );
        Pico_Rel_Pos m = { '!', {10, 10}, PICO_ANCHOR_NW };
        Pico_Rel_Pos w = { '!', {0, 0}, PICO_ANCHOR_NW };
        pico_cv_pos("world", &w, NULL, &m);
        printf("  got (%g,%g)\n", w.x, w.y);
        assert((int)w.x == 20 && (int)w.y == 20);
    }

    printf("\n=== ALL TESTS PASSED ===\n");

    pico_init(0);
    return 0;
}

File: lua/tst/cv.lua
pico.init(true)
pico.set.scene { dim={'!', w=100, h=100} }

print "pico.cv.pos.to / pos.from: world <-> window"
do
    -- world 100x100, window 500x500 -> 5x scale
    print('', 1)
    local out = {'!', anchor='NW'}
    pico.cv.pos.to(out, {'!', x=50, y=50, anchor='NW'}, 'window')
    assert(out.x==250 and out.y==250)

    print('', 2)
    local out = {'!', anchor='NW'}
    pico.cv.pos.from(out, {'!', x=250, y=250, anchor='NW'}, 'window')
    assert(out.x==50 and out.y==50)

    print('', 3)
    local win = {'!', anchor='NW'}
    pico.cv.pos.to(win, {'!', x=37, y=89, anchor='NW'}, 'window')
    local back = {'!', anchor='NW'}
    pico.cv.pos.from(back, win, 'window')
    assert(back.x==37 and back.y==89)
end

print "pico.cv.* sub-layer 2-hop walk"
do
    pico.layer.empty('world', 'sub_cv', true, {'!', w=50, h=50})
    pico.set.layer 'sub_cv'

    print('', 'pos sub -> world')
    local w = {'!', anchor='NW'}
    pico.cv.pos.to(w, {'!', x=10, y=20, anchor='NW'}, 'world')
    assert(w.x==20 and w.y==40)

    print('', 'pos world -> sub')
    local s = {'!', anchor='NW'}
    pico.cv.pos.from(s, {'!', x=20, y=40, anchor='NW'}, 'world')
    assert(s.x==10 and s.y==20)

    print('', 'pos sub -> window (2 hops)')
    local win = {'!', anchor='NW'}
    pico.cv.pos.to(win, {'!', x=10, y=20, anchor='NW'}, 'window')
    assert(win.x==100 and win.y==200)

    print('', 'rect sub -> world')
    local w = {'!', anchor='NW'}
    pico.cv.rect.to(w, {'!', x=10, y=20, w=5, h=10, anchor='NW'}, 'world')
    assert(w.x==20 and w.y==40 and w.w==10 and w.h==20)

    print('', 'dim sub -> world')
    local v = pico.cv.dim.to('!', {'!', w=5, h=10}, 'world')
    assert(v.w==10 and v.h==20)

    pico.set.layer 'world'
end

print "pico.cv.* descendant target/source (bidirectional)"
do
    -- cur=world, target=sub_cv (descendant). sub_cv is 50x50 in world 100x100.
    print('', 'pos world -> sub (target=descendant)')
    local s = {'!', anchor='NW'}
    pico.cv.pos.to(s, {'!', x=20, y=40, anchor='NW'}, 'sub_cv')
    assert(s.x==10 and s.y==20)

    print('', 'pos sub -> world (source=descendant)')
    local w = {'!', anchor='NW'}
    pico.cv.pos.from(w, {'!', x=10, y=20, anchor='NW'}, 'sub_cv')
    assert(w.x==20 and w.y==40)

    print('', 'rect world -> sub (target=descendant)')
    local s = {'!', anchor='NW'}
    pico.cv.rect.to(s, {'!', x=20, y=40, w=10, h=20, anchor='NW'}, 'sub_cv')
    assert(s.x==10 and s.y==20 and s.w==5 and s.h==10)

    print('', 'dim world -> sub (target=descendant)')
    local v = pico.cv.dim.to('!', {'!', w=10, h=20}, 'sub_cv')
    assert(v.w==5 and v.h==10)
end

print "pico.cv.* mode-string as to (returns fresh table)"
do
    -- mode-string in the to slot returns a fresh table; no layer = cur.
    print('', "pos.to with '!'")
    local out = pico.cv.pos.to('!', {'!', x=50, y=50, anchor='NW'}, 'window')
    assert(out and out.x==250 and out.y==250 and out[1]=='!')

    print('', "rect.from with '%'")
    local out = pico.cv.rect.from('%', {'!', x=250, y=250, w=50, h=50, anchor='NW'}, 'window')
    assert(out and out[1]=='%')

    print('', "dim.to with '#'")
    local out = pico.cv.dim.to('#', {'!', w=10, h=20}, 'window')
    assert(out and out[1]=='#')

    print('', "invalid mode-string 'z' errors")
    assert(not pcall(pico.cv.pos.to, 'z', {'!', x=0, y=0, anchor='NW'}, 'window'))
end

print "\n=== ALL TESTS PASSED ==="
pico.init(false)
